lib/portlayer/exec/container.go

```diff
@@ -451,8 +452,21 @@ func (c *Container) Remove(ctx context.Context, sess *session.Session) error {
 		return c.vm.DeleteExceptDisks(ctx)
 	})
 	if err != nil {
-		c.State = existingState
-		return err
+		f, ok := err.(types.HasFault)
+		if !ok {
+			c.State = existingState
+			return err
+		}
+		switch f.Fault().(type) {
+		case *types.InvalidState:
+			log.Warnf("container VM is in invalid state, unregistering")
+			if err := c.vm.Unregister(ctx); err != nil {
+				log.Errorf("Error while attempting to unregister container VM: %s", err)
+				return err
```

Comment:
I don't think we should reset the state. We don't currently know at which point in DeleteExceptDisks we failed. It's possible the VM doesn't have a root disk any more if they were successfully detached.
The case we're attempting to handle (InvalidState due to guestinfo update during shutdown) should mean that the failure is all or nothing, but unless we rework DeleteExceptDisks to allow us to be sure of that we shouldn't assume it here.
Stick with the current Removing state.